design/open/uncomitted-gc.md

```diff
@@ -0,0 +1,119 @@
+# Uncommitted Garbage Collector
+
+## Motivation
+
+Uncommitted data which is no longer referenced (due to branch deletion, reset branch etc.) is not being deleted by lakeFS.
+This may result excessive storage usage and possible compliance issues.
+To solve this problem two approaches were suggested:
+1. A batch operation performed as part of an external process (GC)
+2. An online solution inside lakeFS
+
+Several attempts for an online solution have been made, most of which are documented [here](Link-to-hard-delete-proposal).
+This document will describe the **offline** GC process for uncommitted objects.
+
+Garbage collection of uncommitted data will be performed using the same principals of the current GC process.
+The basis for this is a GC Client (i.e. _Spark_ job) consuming objects information from both lakeFS and the storage adapter, 
+and using this information to determine which objects on the bucket can be deleted.
+
+For simplicity, in this document we will reduce this problem to the repository and branch level.
+
+## Design
+
+### Required changes by lakeFS
+
+The following are necessary changes in lakeFS in order to implement this proposal successfully.
+
+#### Objects Path Conventions
+
+1. Repository objects will be stored under the prefix `<storage_namespace>/repos/<repo_uid>/`
+2. For each repository branch, objects will be stored under the repo prefix with the path `branches/<branch_id>/`
+3. Each lakeFS instance will create a unique prefix partition under the branch path. This will be used by the specific
+lakeFS instance to store the branch's objects. This prefix will be composed of two parts:
+   1. Lexicographically sortable, descending time based serialization
+   2. A unique identifier for the lakeFS instance
+      `<sortable_descending_serialized_uid>_<lakefs_instance_uid>`  
+This serialized partition prefix will allow partial scans of the bucket when running the optimized GC
+4. The lakeFS instance will track the count of objects uploaded to the prefix and create a new partition every < TBD > objects uploaded
+
+The full object path for an object `file1` uploaded to repository `my-repo` on branch `test-branch` will be as follows:  
+    
+    <lakeFS_bucket>/repos/my-repo/branches/test-branch/<lakeFS_instance_partition_uid>/<file1_uid>
+
+#### StageObject
+
+1. Allowed only for address outside the repo namespace
+2. Prevent race between staging object and GC job
```

Comment:
What kind of race? Is this the token thing?

Reply:
By not allowing StageObject on paths inside the repo namespace, we prevent a race where an un-referenced address is being staged using StageObject while it is being deleted by GC